src/System.Private.ServiceModel/tests/Common/Unit/TestHelpers.cs

```diff
@@ -276,3 +279,89 @@ protected override void OnWriteBodyContents(XmlDictionaryWriter writer)
         writer.WriteString(_bodyContent);
     }
 }
+
+public class WcfDuplexServiceCallback : IWcfDuplexServiceCallback
+{
+    public Task<Guid> OnPingCallback(Guid guid)
+    {
+        return Task.FromResult<Guid>(guid);
+    }
+
+    void IWcfDuplexServiceCallback.OnPingCallback(Guid guid)
+    {
+
+    }
+}
+
+public class MyOperationBehavior : Attribute, IOperationBehavior
+{
+    public static StringBuilder errorBuilder = new StringBuilder();
+    public static ManualResetEvent mre = new ManualResetEvent(false);
+    public bool validateMethod = false;
+    public bool addBindingParametersMethod = false;
+    public bool applyClientBehaviorMethod = false;
+    public string errorMessage = "method was called out-of-order, the correct order is: Validate, AddBindingParameters, ApplyClientBehavior.";
+
+    // Pass data at runtime to bindings to support custom behavior
+    public void AddBindingParameters(OperationDescription operationDescription, BindingParameterCollection bindingParameters)
+    {
+        addBindingParametersMethod = true;
+        if ((validateMethod == false) || (applyClientBehaviorMethod))
+        {
+            errorBuilder.AppendLine(String.Format("AddBindingParameters {1}", errorMessage));
+        }
+
+        if (operationDescription == null)
+        {
+            errorBuilder.AppendLine(String.Format("A parameter passed into the AddBindingParameters method was null/nThe null parameter is: {0}", typeof(OperationDescription).ToString()));
+        }
+        if (bindingParameters == null)
+        {
+            errorBuilder.AppendLine(String.Format("A parameter passed into the AddBindingParameters method was null/nThe null parameter is: {0}", typeof(BindingParameterCollection).ToString()));
+        }
+    }
+
+    // Implements a modification or extension of the client across an operation
+    public void ApplyClientBehavior(OperationDescription operationDescription, ClientOperation clientOperation)
+    {
+        applyClientBehaviorMethod = true;
+        if ((validateMethod == false) || (addBindingParametersMethod == false))
+        {
+            errorBuilder.AppendLine(String.Format("ApplyClientBehavior {1}", errorMessage));
+        }
+        else
+        {
+            mre.Set();
+        }
```

Comment:
I'd recommend doing the other error checking before the Set, otherwise your code to test this might be able to execute before you finished with the code below the Set.  In other words, I'd do all the errorBuilder stuff first.